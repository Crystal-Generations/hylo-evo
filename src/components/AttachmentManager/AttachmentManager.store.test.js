/* eslint-env jest */
import reducer, {
  SET_ATTACHMENTS,
  ADD_ATTACHMENT,
  REMOVE_ATTACHMENT,
  SWITCH_ATTACHMENTS,
  defaultState
} from './AttachmentManager.store'

describe('AttachmentManager store', () => {
  describe('reducer', () => {
    describe('when SET_ATTACHMENTS', () => {
      const action = {
        type: SET_ATTACHMENTS,
        payload: {
          polymorphicId: 'post-new',
          attachmentType: 'image',
          attachments: [
            'foo.png',
            'bar.jog'
          ]
        }
      }
      const action2 = {
        type: SET_ATTACHMENTS,
        payload: {
          polymorphicId: 'post-new',
          attachmentType: 'file',
          attachments: [
            'foo.pdf',
            'bar.xls'
          ]
        }
      }
      it('sets attachments', () => {
        let finalState = reducer(reducer(defaultState, action), action2)
        expect(finalState).toMatchSnapshot()
      })
    })

    describe('when ADD_ATTACHMENT', () => {
      const action = {
        type: ADD_ATTACHMENT,
        payload: {
          polymorphicId: 'post-new',
          attachmentType: 'file',
          url: 'bar.pdf',
          position: 0
        }
      }
      it('adds an attachment', () => {
        const finalState = reducer(defaultState, action)
        expect(finalState).toMatchSnapshot()
      })
    })

    describe('when REMOVE_ATTACHMENT', () => {
      const action = {
        type: REMOVE_ATTACHMENT,
        payload: {
          polymorphicId: 'post-new',
          attachmentType: 'file',
          position: 1
        }
      }
      it('removes an attachment', () => {
        const state = {
          ...defaultState,
          ['post-new']: {
            ['image']: [
              'a.png',
              'b.png',
              'c.png'  
            ],
            ['file']: [
              'a.pdf',
              'b.pdf',
              'c.pdf'  
            ]
          }
        }
        const finalState = reducer(state, action)
        expect(finalState).toMatchSnapshot()
      })
    })

    describe('when SWITCH_ATTACHMENTS', () => {
      const action = {
        type: SWITCH_ATTACHMENTS,
        payload: {
          polymorphicId: 'post-new',
          attachmentType: 'image',
          position1: 1,
          position2: 2,
        }
      }
      it('switches two attachments', () => {
        const state = {
          ...defaultState,
          ['post-new']: {
            ['image']: [
              'a.png',
              'b.png',
              'c.png'  
            ],
            ['file']: [
              'a.pdf',
              'b.pdf',
              'c.pdf'  
            ]
          }
        }
        const finalState = reducer(state, action)
        expect(finalState).toMatchSnapshot()
      })
    })
  })

  describe('upload', () => {
    const url = 'http://filepicker.io/hfwoe/eh98e'
    const filename = 'foo.jpg'
    const opts = { type: 'userAvatar', id: 3, attachmentType: 'image' }
    
    it('returns an API action after filepicker succeeds', () => {
      const uploadAttachment = setupUploadAttachment(true, [url, filename])
      const action = uploadAttachment(opts)
    
      expect(action).toEqual({
        type: 'UPLOAD_ATTACHMENT',
        payload: expect.any(Promise),
        meta: opts
      })
    
      return expect(action.payload).resolves.toEqual({
        api: {
          method: 'post',
          path: '/noo/upload',
          params: { url, id: 3, type: 'userAvatar', filename }
        }
      })
    })
    
    it('rejects when picker fails', () => {
      const uploadAttachment = setupUploadAttachment(false, [new Error('nope')])
      return expect(uploadAttachment(opts).payload).rejects.toEqual(new Error('nope'))
    })
  })
})

function setupUploadAttachment (shouldSucceed, callbackArgs) {
  jest.resetModules()

  jest.doMock('client/filepicker', () => {
    if (shouldSucceed) {
      return { uploadFile: opts => opts.success(...callbackArgs) }
    } else {
      return { uploadFile: opts => opts.failure(...callbackArgs) }
    }
  })

  const { uploadAttachment } = require('./AttachmentManager.store')
  return uploadAttachment
}
