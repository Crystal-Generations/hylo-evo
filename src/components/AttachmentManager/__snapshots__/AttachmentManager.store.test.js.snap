// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentManager store reducer when ADD_ATTACHMENT adds an attachment 1`] = `
Object {
  "post-new-file": Array [
    Object {
      "url": "bar.pdf",
    },
  ],
}
`;

exports[`AttachmentManager store reducer when REMOVE_ATTACHMENT removes an attachment 1`] = `
Object {
  "post-new-file": Array [
    "a.pdf",
    "c.pdf",
  ],
  "post-new-image": Array [
    "a.png",
    "b.png",
    "c.png",
  ],
}
`;

exports[`AttachmentManager store reducer when SET_ATTACHMENTS sets attachments 1`] = `
Object {
  "post-new-file": Array [
    "foo.pdf",
    "bar.xls",
  ],
  "post-new-image": Array [
    "foo.png",
    "bar.jog",
  ],
}
`;

exports[`AttachmentManager store reducer when SWITCH_ATTACHMENTS switches two attachments 1`] = `
Object {
  "post-new-file": Array [
    "a.pdf",
    "b.pdf",
    "c.pdf",
  ],
  "post-new-image": Array [
    "a.png",
    "c.png",
    "b.png",
  ],
}
`;
